src/protocol/resp/src/parse/command.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright 2023 Pelikan Foundation LLC.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use std::borrow::Cow;
+use std::str::FromStr;
+
+use crate::message::Message;
+
+use super::{ArrayParser, ParseError, ParseResult, Parser};
+
+enum Detail<'a, 'p> {
+    Inline(Vec<Cow<'a, [u8]>>),
+    Resp(ArrayParser<'a, 'p>),
+}
+
+pub struct CommandParser<'a, 'p> {
+    detail: Detail<'a, 'p>,
+}
+
+impl<'a, 'p> CommandParser<'a, 'p> {
+    pub fn new(parser: &'p mut Parser<'a>) -> ParseResult<'a, Self> {
+        match parser.peek() {
+            Some(b'+' | b'-' | b':' | b'$' | b'*') => Ok(Self {
```

Comment:
Shouldn't the leading character always be `*` for RESP, non-inlined command?
nit: As a convenience, add https://redis.io/docs/reference/protocol-spec/ as a reference to the implementation documentation.

Reply:
This was copied from the existing parser. A valid RESP command will always start with `*` but I think the reason was to avoid parsing a valid redis protocol object as an inline command. (And also there's was something about custom commands that aren't arrays?)

Reply:
I acknowledge this is not a behavior you introduced, but it is the first time I noticed 😛 

>A client sends the Redis server a RESP Array consisting of only Bulk Strings.
>A Redis server replies to clients, sending any valid RESP data type as a reply.

I am not aware of valid non-array commands, even though I complained about only allowing _Bulk String_ in the command which is always an array (e.g. allowing maps). 

All inline commands start with a verb which iirc always starts with a `[a-zA-Z]` character. I don't think we need match the allowed letters exactly at this step, but if we draw the line at RESP command versus non RESP command,  then there's only one permitted first character for the former.

Reply:
as a compromise, we could put the only presumed valid character first in the match so the branch prediction almost always succeeds (assuming that Rust compiles the match statement in this case to the most desirable switch statement type assembly) until we decide if we want to split the parsing so that the incoming command parsing is as strict as Redis allows (my interpretation of @thinkingfish request)

in this example: https://godbolt.org/z/ca4fcnerE

I have * first and it appears to compare it first here:

``` asm

example::match_char:
        sub     rsp, 72
        mov     al, sil
        mov     cl, dil
        and     cl, 1
        mov     byte ptr [rsp + 16], cl
        mov     byte ptr [rsp + 17], al
        mov     al, byte ptr [rsp + 16]
        and     al, 1
        movzx   eax, al
        cmp     rax, 1
        jne     .LBB1_2
        mov     al, byte ptr [rsp + 17]
        mov     byte ptr [rsp + 15], al
        sub     al, 42
        je      .LBB1_3
        jmp     .LBB1_4
.LBB1_4:
        mov     al, byte ptr [rsp + 15]
        sub     al, 45
        je      .LBB1_3
        jmp     .LBB1_2
```

with the following:

```
    sub     al, 42
    je      .LBB1_3
    jmp     .LBB1_4
```

comparing 42 in ASCII, which is '*', before jumping if not equal to `.LBBi_4`, which compares against the number `45` which is '-' in ASCII

so I think performance-wise, moving '*' would be desirable, but wouldn't do anything to make the parser more restrictive

Reply:
I think changing it to allow only arrays is reasonable.

Reply:
Some responses to points above:
- This code _already_ only supports arrays. With the way things are here the error people get if they do that is along the lines of "expected an array, got a string/int/whatever". If I remove the extra bits here the the error message will be something like `"unknown command '$-1'"` which is far worse IMO.
- As for performance, this code is only run once per request. Trying to micro-optimize a match statement that is run once per request is not worth the hit to readability.

I don't think there's any changes needed here since it already only restricts itself to valid commands and changing things up would result in worse error messages to the user.

Reply:
@swlynch99 why would rearranging the order of the characters hurt readability? not arguing we should do it, just curious

Reply:
Oh, sorry, misread your comment. Yeah that would not hurt readability at all. Let me do that now

Reply:
@hderms Coming back to this. A comparison with optimized codegen is probably more useful: https://godbolt.org/z/bajq6bdrT. Godbolt doesn't turn that on by default

Reply:
@swlynch99 oh yeah, that's much easier to read

Reply:
@swlynch99 I don't understand ASM very well but it looks like it's using that huge number as a bit string which it somehow encodes the rest of the comparisons into?

Reply:
Yeah exactly. That number has bits set for the ascii values of the character and `bt` extracts a bit and puts it in the `CF` flag.